File: server/src/external/images.ts
import axios from 'axios'
import fs from 'fs'
import path from 'path'
import sharp from 'sharp'
import type { MovieTypeTMDB } from '../movie-type.js'
import { toFilename } from '../utils.js'

const IMAGES_DIR = path.resolve(process.cwd(), './public/images')

export const getTmdbImage = async (
    movie: MovieTypeTMDB,
    replace = false,
): Promise<{ error: string } | { path: string }> => {
    if (!movie.tmdbPosterPath) {
        return { error: 'No poster path' }
    }

    try {
        const url = `https://image.tmdb.org/t/p/w500/${movie.tmdbPosterPath}`
        const filename = toFilename(movie.name)
        const localPath = `${IMAGES_DIR}/${filename}.webp`
        const frontendPath = `images/${filename}.webp`

        if (fs.existsSync(localPath) && !replace) {
            console.log('File exists, skipping:', localPath)
            return { path: frontendPath }
        }

        const response = await axios({
            url,
            method: 'GET',
            responseType: 'arraybuffer',
        })

        const imgData = await sharp(response.data)
            .resize({
                width: 400,
                withoutEnlargement: true,
            })
            .webp({ effort: 5, quality: 70 })
            .toBuffer()

        await fs.promises.writeFile(localPath, imgData)
        console.log(`   🖼️ Downloaded poster for ${movie.name}`)
        return { path: frontendPath }
    } catch (e: unknown) {
        console.log('   🛑 Error fetching poster for', movie.name, e)
        return { error: 'error fetching poster' }
    }
}

File: api/app/external/images.py
import os
from io import BytesIO

import httpx
from PIL import Image

from app.config import settings
from app.exceptions import ImageError
from app.utils import to_filename


async def download_poster(name: str, tmdb_poster_path: str, *, replace: bool = False):
    filename = to_filename(name)

    tmdb_url = f"https://image.tmdb.org/t/p/w500/{tmdb_poster_path}"
    local_path = f"{settings.images_dir}/{filename}.webp"
    frontend_path = f"/images/{filename}.webp"

    if os.path.exists(local_path) and not replace:
        return frontend_path

    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(tmdb_url)
            response.raise_for_status()
    except httpx.HTTPError as e:
        raise ImageError(f"poster download failed for {name}: {e}") from e

    try:
        img = Image.open(BytesIO(response.content))
        img.thumbnail((500, 750))
        img.save(local_path, "WEBP", quality=70)
    except (OSError, ValueError) as e:
        raise ImageError(f"poster resize failed for {name!r}: {e}") from e

    return frontend_path
